Fetch Marketplace promotions - ensure callback is always registered for `woocommerce_marketplace_fetch_promotions` action

mcliwanow opened the pull request:
We started getting errors 'Scheduled action for woocommerce_marketplace_fetch_promotions will not be executed as no callbacks are registered.' on some environments.

In this commit we ensure that WC_Admin_Marketplace_Promotions::fetch_marketplace_promotions() is always added as a calback to our action hook.

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

1. Ensure that file `plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php` is always loaded. This is to ensure that callback for previously scheduled action (added in WC Core 8.8) is always there.
2. Change how we fetch Marketplace promotions by switching from Action Scheduler to transients. This makes it easier to ensure that action is called only in admin context and has no impact on frontend users.
3. Reuse the `woocommerce_marketplace_suppress_promotions` filter in more places to ensure that it suppresses fetching data and displaying it.

Closes #46923 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. Check out this branch.
2. Apply the following patch (it adds 2 `error_log()` entries and changes URL for fetching promotions) 
[promotions_transients.patch](https://github.com/woocommerce/woocommerce/files/15317832/promotions_transients.patch)

3. Build WooCommerce
```
cd plugins/woocommerce
nvm use
pnpm install
pnpm --filter='@woocommerce/plugin-woocommerce' build
```
4. Enter docker running at port 8888 (`wordpress-1`) and tail debug logs
```
cd wp-content
tail -f debug.log
```
5. Enter CLI docker (Name: `cli-1`) separately
6. Use the site as a regular user
   - confirm that there are no new log entries
   - Ensure that transient is not set - `wp transient get woocommerce_marketplace_promotions_v2` should be empty 
8. View site as admin
   - confirm there's one log for `fetch promotions`
   - browse through admin pages and confirm that remaining logs are `skip fetching promotions`
   - Go back to CLI docker and confirm that there's a value for the transient `wp transient get woocommerce_marketplace_promotions_v2`
9. Confirm that you see Sale bubble in the menu and that Sale notification is displayed in the In-App Marketplace
![image](https://github.com/woocommerce/woocommerce/assets/4765119/ec2b45d8-763d-4220-9ea1-e977483df45e)
10. In CLI docker enter wp shell by running `wp shell` and run `WC_Admin_Marketplace_Promotions::get_active_promotions();` - confirm that active promotions return only what is matching current dates
11. Still in WP shell, remove the transient `delete_transient('woocommerce_marketplace_promotions_v2');`
12. Refresh admin pages and ensure that log `fetch promotions` is there
13. Add filter `add_filter( 'woocommerce_marketplace_suppress_promotions', '__return_true' );` and ensure that both sale bubble and notification inside In-App are no longer visible
14. Go to http://wccore.test:8888/wp-admin/admin.php?page=wc-status&tab=action-scheduler&action=-1&action2=-1&status=pending
2. Check if you have `woocommerce_marketplace_fetch_promotions` in pending actions
3. If yes, run the action
4. Make sure that afterward action `woocommerce_marketplace_fetch_promotions_clear` runs too and that when it's done you see neither `woocommerce_marketplace_fetch_promotions` nor `woocommerce_marketplace_fetch_promotions_clear` in pending actions
15. Please test around this issue / Try to break things

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

mcliwanow committed:
Marketplace promotions - ensure action is always there

We started getting errors 'Scheduled action for woocommerce_marketplace_fetch_promotions will not be executed as no callbacks are registered.' on some environments.

In this commit we ensure that WC_Admin_Marketplace_Promotions::fetch_marketplace_promotions() is always added as a calback to our action hook.

mcliwanow committed:
Changelog

andfinally commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -585,6 +585,7 @@ public function includes() {
 		include_once WC_ABSPATH . 'includes/queue/class-wc-action-queue.php';
 		include_once WC_ABSPATH . 'includes/queue/class-wc-queue.php';
 		include_once WC_ABSPATH . 'includes/admin/marketplace-suggestions/class-wc-marketplace-updater.php';
+		include_once WC_ABSPATH . 'includes/admin/class-wc-admin-marketplace-promotions.php';
```
This means `WC_Admin_Marketplace_Promotions::init` is run for every request, even front end Ajax and Cron requests. It seems a shame that we have to do this simply because some users apparently run cron without setting `DOING_CRON`. But I can see there's not much else we can do to account for that case. And we have enough conditions in `init` to mean this should not be too heavy.

andfinally replied:
This is now superseded by the new method of fetching.

andfinally reviewed (approved):
Thanks Michal, this works as expected. 💯 

Hopefully rebasing on trunk will fix the test errors.

We should probably change these last points in the steps:

> 15. add filter `woocommerce_marketplace_suppress_promotions` that always returns true
> 16. Visit the [pending scheduled actions page](http://localhost:8888/wp-admin/admin.php?page=wc-status&tab=action-scheduler&status=pending) and search for `woocommerce_marketplace_fetch_promotions`. If there's a pending action, cancel it using the "cancel" hover link under it in the list. A new action should be scheduled.
> 17. Ensure that the action is available in the Pending tab and after a while it should be available in new tab Past-due (you might need to refresh the page)
> 18. Run the action via CLI `pnpm -- wp-env run cli wp cron event run action_scheduler_run_queue` and ensure that it's completed successfully - logs should indicate that early exits were triggered properly

because when the filter returns true, we don't even schedule the action. So when you cancel the remaining scheduled action, no new one is created, and the CLI command will do nothing. This is fine!

It was impossible to remove the value of `DISABLE_WP_CRON` once I'd set it – restarting wp-env didn't clear it, even though it was removed from the config file. I eventually had to set it to false.

mcliwanow committed:
Marketplace promotions - switch from Action Scheduler to transients

We're switching fetching marketplace promotions from using Action Scheduler
every 12 hours, to a 1 day transient. This makes it easier to ensure that
the requests are triggered only for admins, and that they don't impact frontend
users of the stores.

mcliwanow committed:
Update changelog

mcliwanow commented:
@andfinally, @corsonr, after some refactoring that changes code from using Action Scheduler to rely on transients (and running code only in wp admin context), I made changes to PR description and testing instructions. Would you mind doing another review?

mcliwanow committed:
Fix lint

andfinally commented on plugins/woocommerce/changelog/fix-woocommerce_marketplace_fetch_promotions-broken-with-cron:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: fix
+
+Switch marketplace promotions from Action Scheduler to cron.
```
We should change this to something like `from Action Scheduler to transient`.

andfinally commented on plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php:
```diff
@@ -38,25 +37,8 @@ class WC_Admin_Marketplace_Promotions {
-			add_action( 'init', array( __CLASS__, 'schedule_promotion_fetch' ), 12 );
-		}
+		// A legacy hook that can be triggered by action scheduler.
+		add_action( 'woocommerce_marketplace_fetch_promotions', '__return_true' );
```
Clever, I'm glad you thought of that!

mcliwanow replied:
Thanks! I've changed it a bit in https://github.com/woocommerce/woocommerce/pull/47262/commits/ddd1cb793a69c6cdeeedc5806b2a27c89d585a2f and I wonder if you wouldn't mind reviewing this part?

I'll add the following steps to testing instructions:
1. Go to http://wccore.test:8888/wp-admin/admin.php?page=wc-status&tab=action-scheduler&action=-1&action2=-1&status=pending
2. Check if you have `woocommerce_marketplace_fetch_promotions` in pending actions
3. If yes, run the action
4. Make sure that afterward action `woocommerce_marketplace_fetch_promotions_clear` runs too and that when it's done you see neither `woocommerce_marketplace_fetch_promotions` nor `woocommerce_marketplace_fetch_promotions_clear` in pending actions

andfinally commented on plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php:
```diff
@@ -292,13 +325,14 @@ private static function append_bubble( string $menu_item_text, string $bubble_te
 
 	/**
-	 * When WooCommerce is deactivated, clear the scheduled action.
+	 * Clear event that was used for fetching promotions in WC Core 8.8.
```
```suggestion
	 * Clear the scheduled action that was used to fetch promotions in WooCommerce 8.8.
```

andfinally reviewed (commented):
This looks great, thanks, and tests as expected. I could have saved us a lot of work by just taking your original suggestion and not the advice of the clever people who recommended scheduled actions!

I had to comment out these options in `fetch_marketplace_promotions`, otherwise the function got a 404 when trying to fetch the gist.

```php
$fetch_options  = array(
	'auth'    => true,
	'country' => true,
);
```


I love the safeguard `add_action` for the legacy hook in `init`. I wonder if it would make sense to go further and cancel the leftover scheduled action? I still have the `woocommerce_marketplace_fetch_promotions` SA scheduled for every 12 hours.

One fairly safe way would be to restore

```php
register_deactivation_hook( WC_PLUGIN_FILE, array( __CLASS__, 'clear_scheduled_event' ) );
```

which would clear it for some people. (We're not using `clear_scheduled_event` otherwise.) We could also add a `register_activation_hook`, as that may be called when the user upgrades WooCommerce. Or we could even check the WooCommerce version on `plugins_loaded`, like discussed [here](https://wordpress.stackexchange.com/a/144873/14984).

mcliwanow committed:
Fix typos

mcliwanow commented:
@andfinally , nice catch with 
```
register_deactivation_hook( WC_PLUGIN_FILE, array( __CLASS__, 'clear_scheduled_event' ) );
```

I wanted to keep it but must have removed it by accident 🤦 

Now I'm thinking about replacing this line `add_action( 'woocommerce_marketplace_fetch_promotions', '__return_true' );` to call a simple callback which would unscheduled the action if it was scheduled. It means that it should run only once then. I'll test it and will update the PR if it works.


> I had to comment out these options in fetch_marketplace_promotions, otherwise the function got a 404 when trying to fetch the gist.

Interesting 🤔 It worked for me without this particular change, but it's a good tip for other reviewers.

mcliwanow committed:
Clear action from Action Scheduler

Action woocommerce_marketplace_fetch_promotions is no longer used, so we're clearing it when it's triggered.
Since we cannot self-clear the action, we're triggering a new single action when woocommerce_marketplace_fetch_promotions is run.
This new action's sole responsibility is to clear the original action.

andfinally reviewed (approved):
Genius solution to the leftover scheduled actions! 🏆 I retested everything and all worked as expected.

mcliwanow commented:
I did one small change after @andfinally's review - minor lint fix, so I just amended the commit. In history it looks like it happened after @andfinally's review but 99% of the changes were reviewed by him.

ilyasfoo commented on plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php:
```diff
@@ -66,31 +49,73 @@ public static function init() {
+			return;
+		}
+
+		self::maybe_update_promotions();
```
Hey, @mcliwanow! I'm working on [identifying synchronous remote requests to WCCOM](https://github.com/woocommerce/woocommerce/issues/50543). Just wanted to flag that this is running synchronously when viewing `wp-admin`. While this has caching to limit requests, could it be made better by running it in a cron instead?

ilyasfoo replied:
Or alternatively, via ajax request

mcliwanow replied:
@ilyasfoo, we can take a look at this one. Before we do though, do you think it'd make sense to create some abstraction layer making it easier to communicate with WCCOM in a safe, async, and cachable way?

ilyasfoo replied:
Thanks, @mcliwanow!

> do you think it'd make sense to create some abstraction layer making it easier to communicate with WCCOM in a safe, async, and cachable way?

We do have an existing generic class ([DataSourcePoller](https://github.com/woocommerce/woocommerce/blob/0a60c3c5ec9b4830ab7d4e3817354f25bec597c3/plugins/woocommerce/src/Admin/RemoteSpecs/DataSourcePoller.php#L1)) which handles API requests, locale, and caching. However, it doesn't implement async queries out of the box.

Feel free to improve the class if you'd like, I'd be glad to help around with reviews and suggestions!

andfinally replied:
Hi @ilyasfoo 👋, IIRC, we had some [problems](https://woocorerelease.wordpress.com/2024/04/16/code-freeze-exception-8-8-2-2/) previously we tried to update this data using Action Scheduler. What's the background to this?

I read on the linked issue that this type of request can break wp-admin and on the P2 it [says](https://ghidorahp2.wordpress.com/2024/08/08/plan-of-attack-remote-specs-faults/#problem-3-synchronous-requests) we want to rewrite such requests. Are we worried for example that if the endpoint requested no longer exists, this will break the rest of the page?

Looking at the [function concerned](https://github.com/woocommerce/woocommerce/blob/536807ca432c8ff466d3726c7735073cc9008234/plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php#L109-L160), we seem to have adequate error handling, so that in this case the error would be contained.

ilyasfoo replied:
Hey, @andfinally. Thanks for looking into it!

> What's the background to this?

The primary goal of changing it to async is to improve the stability `wp-admin` experience. As of now, the request runs synchronously and runs on all `wp-admin` pages. In the event of an unresponsive `woocommerce.com` and cache not available, it can potentially hold up `wp-admin` requests at least until a cache is set. 

> Are we worried for example that if the endpoint requested no longer exists, this will break the rest of the page?

While it's unlikely for `woocommerce.com` to go down, when it happens, it could stall merchants/admin users even if don't intend to use WooCommerce. In a typical configuration, API requests or PHP execution timeouts can take up to 30 seconds. I think it's still worth attempting to improve the experience.

We can simulate the current issue by using beta tester tool with the following steps:

1. Create a fresh Jurassic Ninja site with `woocommerce`
2. Install `woocommerce-beta-tester` (download [woocommerce-beta-tester.zip](https://github.com/user-attachments/files/17094190/woocommerce-beta-tester.zip) or build from `trunk` in core repo)
3. Go to `Tools > WCA Test Helper > Tools`
4. In `Force errors on woocommerce.com requests`, select `Timeout after 6 seconds` and click `Run` ([screenshot](https://github.com/user-attachments/assets/de3ab026-5bb7-41de-8f5f-c383da9916f1))
5. SSH to the site, and run `wp cache flush`
6. Go to `/wp-admin`
7. Observe the request takes at least 6 seconds to load

andfinally replied:
HI @ilyasfoo thanks for the details! We have an [issue](21774-gh-Automattic/woocommerce.com) to fix this in the upcoming release.
